File: frontend/src/api/recommendation.ts
import api from "./client";
import type { Recommendation, Approval, Execution } from "../types";

export const rulesApi = {
  generate: (date?: string, supplierCode?: string) =>
    api.post<Recommendation[]>("/rules/generate", null, {
      params: { target_date: date, supplier_code: supplierCode },
    }),
};

export const recommendationApi = {
  list: (params?: { target_date?: string; supplier_code?: string; status?: string }) =>
    api.get<Recommendation[]>("/recommendation", { params }),
  approve: (id: number, body: { action: string; approved_qty?: number; memo?: string }) =>
    api.post<Approval>(`/recommendation/${id}/approve`, body),
  execute: (id: number, actual_packed_qty: number) =>
    api.post<Execution>(`/recommendation/${id}/execute`, { actual_packed_qty }),
  updateExecution: (execId: number, body: { used_qty?: number; remaining_qty?: number; status?: string }) =>
    api.patch<Execution>(`/recommendation/execution/${execId}`, body),
};

File: backend/routers/recommendation.py
from datetime import date
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.ext.asyncio import AsyncSession

from database import get_db
from schemas.recommendation import (
    RecommendationOut, ApprovalRequest, ApprovalOut,
    ExecutionRequest, ExecutionUpdateRequest, ExecutionOut,
)
from services.recommendation import (
    get_recommendations, approve_recommendation,
    create_execution, update_execution,
)

router = APIRouter(prefix="/api/recommendation", tags=["추천/승인"])


@router.get("", response_model=list[RecommendationOut])
async def list_recommendations(
    target_date: date = Query(default=None),
    supplier_code: str = Query(default=None),
    status: str = Query(default=None),
    db: AsyncSession = Depends(get_db),
):
    return await get_recommendations(db, target_date, supplier_code, status)


@router.post("/{rec_id}/approve", response_model=ApprovalOut)
async def approve(rec_id: int, body: ApprovalRequest, db: AsyncSession = Depends(get_db)):
    try:
        return await approve_recommendation(db, rec_id, body.action, body.approved_qty, body.memo)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.post("/{rec_id}/execute", response_model=ExecutionOut)
async def execute(rec_id: int, body: ExecutionRequest, db: AsyncSession = Depends(get_db)):
    from sqlalchemy import select
    from models.approval import Approval
    result = await db.execute(select(Approval).where(Approval.recommendation_id == rec_id))
    approval = result.scalars().first()
    if not approval:
        raise HTTPException(status_code=404, detail="승인 정보를 찾을 수 없습니다.")
    try:
        return await create_execution(db, approval.id, body.actual_packed_qty)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.patch("/execution/{exec_id}", response_model=ExecutionOut)
async def update_exec(exec_id: int, body: ExecutionUpdateRequest, db: AsyncSession = Depends(get_db)):
    try:
        return await update_execution(db, exec_id, body.used_qty, body.remaining_qty, body.status)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
